### PR TITLE
[WIP] Add invoices & payments integration into system

### PR DESCRIPTION
**Pull request by coderberry**
These pages are hidden, but not blocked for the publishers. We will add these pages to the left sidebar once we have the data confirmed as accurate.

### CODE REVIEW

File: app/models/invoice.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    Monetize.parse(ad_revenue - ad_spend + bonus_referral + bonus_direct)
+  end
+
+  def paid_amount
+    Monetize.parse(invoice_payment&.amount || 0)
+  end
+
+  def owed_amount
+    total_amount - paid_amount
+  end
```
Comment (hopsoft) [inline app/models/invoice.rb]: An `InvoicePayment` can be applied to multiple invoices. Consider the scenario where it takes a publisher 3 months of earnings to reach the minimum total before pay out. They would have 3 invoices created but only a single payment record. This scenario would produce unexpected behavior for the methods `paid_amount` _(which would be greater than each individual invoice)_ and `owed_amount` _(which would show a negative value after the publisher was paid)_.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -35,6 +37,10 @@ def clear_searches(except: [])
     end
   end
 
+  def layout_by_authentication_status
+    current_user ? "admin" : "application"
+  end
```
[the remark this answers is not in the document]
Comment (coderberry) [reply app/controllers/application_controller.rb]: I used to have it as "authenticated" and "unauthenticated". I agree that it's a bit weird. Let me consider some alternative names, but I don't want to hold up the release for this.

File: app/jobs/deliver_monthly_earnings_report_job.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class DeliverMonthlyEarningsReportJob < ApplicationJob
+  queue_as :default
+
+  def perform(user_id, date)
+    date = Date.coerce(date)
+    user = User.find(user_id)
+    invoice = user.invoices.for_date(date).first
+    total_unpaid_earnings = user.invoices.includes(:invoice_payment).sum(&:owed)
+    return unless invoice
```
[the remark this answers is not in the document]
Comment (coderberry) [reply app/jobs/deliver_monthly_earnings_report_job.rb]: Ah yes. I didn't finish this one.

File: app/models/invoice.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  scope :without_invoice_payments, -> { where(invoice_payment_id: nil) }
+  scope :with_transactions, -> {
+    where.not(ad_revenue_cents: 0).or(
+      Invoice.where(ad_spend_cents: 0).or(
+        Invoice.where(bonus_referral_cents: 0).or(
+          Invoice.where(bonus_direct_cents: 0)
+        )
+      )
+    )
+  }
```
[the remark this answers is not in the document]
Comment (coderberry) [reply app/models/invoice.rb]: love it

File: app/models/invoice.rb (hunk 1 above)
Comment (coderberry) [reply app/models/invoice.rb]: I agree.. this is the one thing that I am a bit concerned with. I think that's why I initially called the table "earnings". It's really just earnings for a full month, where the invoice_payment is more the invoice. It seems this is incomplete.

File: app/controllers/invoice_payments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      .permit(
+        :user_id,
+        :payment_method,
+        :amount,
+        :payment_transaction_id,
+        :paid_by,
+        :details,
+        invoice_ids: [] # virtual attribute used in after save callback
+      ).tap do |whitelisted|
+        whitelisted[:payment_date] = Date.strptime(params[:invoice_payment][:payment_date], "%m/%d/%Y")
```
[the remark this answers is not in the document]
Comment (coderberry) [reply app/controllers/invoice_payments_controller.rb]: +1

**Comment by hopsoft** [comment]
Note that Joseph and I have some proposed _relatively big_ changes around payments. It will be good to circle back on this with him after this months manual payouts.

**Pull request closed without merging**